rpc/RPCSession.java

```diff
@@ -153,44 +155,39 @@ public void run() {
 
     public static class Cluster implements Grakn.Session {
         private static final Logger LOG = LoggerFactory.getLogger(Cluster.class);
+        public static final int MAX_RETRY_PER_SERVER = 3;
         private final GraknClient.Cluster clusterClient;
-        private final Database database;
+        private Database database;
         private final String dbName;
         private final Type type;
-        private final GraknOptions options;
-        private final ConcurrentMap<Replica.Id, RPCSession.Core> coreSessions;
-        private GraknClusterGrpc.GraknClusterBlockingStub clusterDiscoveryRPC;
+        private final GraknOptions.Cluster options;
+        private final ConcurrentMap<Replica.Id, Core> coreSessions;
         private boolean isOpen;
 
-        public Cluster(GraknClient.Cluster clusterClient, String database, Grakn.Session.Type type, GraknOptions options, GraknClusterGrpc.GraknClusterBlockingStub clusterDiscoveryRPC) {
+        public Cluster(GraknClient.Cluster clusterClient, String database, Grakn.Session.Type type, GraknOptions.Cluster options) {
             this.clusterClient = clusterClient;
             this.dbName = database;
             this.type = type;
             this.options = options;
-            this.clusterDiscoveryRPC = clusterDiscoveryRPC;
-            this.database = discoverDatabaseReplicas();
+            this.database = databaseDiscover();
             coreSessions = new ConcurrentHashMap<>();
             isOpen = true;
         }
 
         @Override
         public Grakn.Transaction transaction(Grakn.Transaction.Type type) {
-            return transaction(type, new GraknOptions());
+            return transaction(type, GraknOptions.cluster());
         }
 
         @Override
         public Grakn.Transaction transaction(Grakn.Transaction.Type type, GraknOptions options) {
-            Replica selected = type != Grakn.Transaction.Type.READ_REPLICA ?
-                    database.selectLeader() : database.selectedReplica();
-            RPCSession.Core selection = coreSessions.computeIfAbsent(
-                    selected.id(),
-                    key -> {
-                        LOG.info("Opening a session to leader '{}'", selected);
-                        return clusterClient.coreClients().get(key.address()).session(key.database(), this.type, this.options);
-                    }
-            );
-            LOG.info("Opening a transaction of of type '{}' to leader '{}'", type, selected);
-            return selection.transaction(type, options);
+            if (!options.isCluster()) throw new GraknClientException(ILLEGAL_CAST, options);
```

Comment:
I believe this line is redundant and should be deleted. `GraknOptions.asCluster` already throws if you're performing an illegal cast.
```suggestion
```

Reply:
Fixed in https://github.com/graknlabs/client-java/pull/243/commits/8f7cf33539175c537a906d305bdf68b4ff3de612